signac/contrib/migration/__init__.py

```diff
@@ -5,102 +5,141 @@
 
 import os
 import sys
-from contextlib import contextmanager
 
 from filelock import FileLock
 from packaging import version
 
-from ...common.config import get_config
 from ...version import SCHEMA_VERSION, __version__
-from .v0_to_v1 import migrate_v0_to_v1
+from .v0_to_v1 import _load_config_v1, _migrate_v0_to_v1
 
 FN_MIGRATION_LOCKFILE = ".SIGNAC_PROJECT_MIGRATION_LOCK"
 
 
-MIGRATIONS = {
-    ("0", "1"): migrate_v0_to_v1,
+# Config loaders must be functions with the signature
+# def config_loader(root_directory: str) -> MutableMapping
+# When a new schema version is introduced, a corresponding loader only needs to
+# be added if the old loader will no longer function.  This dictionary must
+# contain all unique loaders for schema versions that are supported as starting
+# points for migration. The resulting MutableMapping config objects must be
+# writeable, i.e. it must be possible to persist in-memory changes from these
+# objects to the underlying config files.
+_CONFIG_LOADERS = {
+    "1": _load_config_v1,
 }
 
 
-def _reload_project_config(project):
-    project_reloaded = project.get_project(
-        root=project.root_directory(), search=False, _ignore_schema_version=True
-    )
-    project._config = project_reloaded._config
+_MIGRATIONS = {
+    ("0", "1"): _migrate_v0_to_v1,
+}
 
 
-def _update_project_config(project, **kwargs):
-    """Update the project configuration."""
-    for fn in ("signac.rc", ".signacrc"):
-        config = get_config(project.fn(fn))
-        if "project" in config:
-            break
-    else:
-        raise RuntimeError("Unable to determine project configuration file.")
-    config.update(kwargs)
-    config.write()
-    _reload_project_config(project)
+_PARSED_SCHEMA_VERSION = version.parse(SCHEMA_VERSION)
```

Comment:
I'm afraid that the term "parsed schema version" originally threw me off, because I thought it was parsed from the directory or so. Maybe just rename it to `_THIS_SCHEMA_VERSION` or so?

Reply:
The intent here and below was to communicate the difference between version strings and `packaging.version.Version` objects. I don't know if `_THIS_SCHEMA_VERSION` really conveys that. Do you have an alternate name that might help with that? Otherwise I can go with your suggestion, but it seems too ambiguous to meaningful differentiate from `SCHEMA_VERSION`.

Reply:
> The intent here and below was to communicate the difference between version strings and packaging.version.Version objects.

I got that, I am just letting you know that I was thrown off by the prefix "parsed_" until I saw that. If you don't like "this_", just leave it _as is_.